modules/schema/src/main/java/org/apache/ignite/internal/util/ColocationUtils.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.BitSet;
+import java.util.UUID;
+import org.apache.ignite.internal.schema.Column;
+import org.apache.ignite.internal.schema.NativeTypeSpec;
+import org.apache.ignite.internal.schema.row.Row;
+
+/**
+ * Colocation hash utilities.
+ */
+public class ColocationUtils {
+    /**
+     * Disallow to construct instance.
+     */
+    private ColocationUtils() {
+        // No-op.
+    }
+
+    /**
+     * Calculate colocation hash of the specified row.
+     *
+     * @param r Row to calculate row's colocation hash.
+     * @return Row's colocation hash.
+     */
+    public static int colocationHash(Row r) {
+        HashCalculator hashCalc = new HashCalculator();
+
+        for (Column c : r.schema().colocationColumns()) {
+            append(hashCalc, r.value(c.schemaIndex()), c.type().spec());
+        }
+
+        return hashCalc.hash();
+    }
+
+
+    /**
+     * Append value to the hash calculation.
+     *
+     * @param calc Hash calculator.
+     * @param v Value to update hash.
+     * @param typeSpec Value type.
+     */
+    public static void append(HashCalculator calc, Object v, NativeTypeSpec typeSpec) {
+        if (v == null) {
+            calc.appendLong(0);
```

Comment:
Will it be better to make calculator responsible for null values hash?
```suggestion
            calc.appendNull();
```